tests/metrics_test.go

```diff
@@ -153,14 +159,119 @@ var _ = Describe("Metrics", func() {
 		)
 	})
 
+	Context("SSP metrics", func() {
+		var originalVersion string
+
+		getDeployment := func(name string, namespace string) *apps.Deployment {
+			deployment := &apps.Deployment{}
+			err := apiClient.Get(ctx, client.ObjectKey{Name: name, Namespace: namespace}, deployment)
+			Expect(err).ToNot(HaveOccurred())
+			return deployment
+		}
+
+		totalRestoredTemplatesCount := func() int {
+			sum := 0
+			operatorPods, operatorMetricsPort := operatorPodsWithMetricsPort()
+			for _, sspOperator := range operatorPods {
+				sum += intMetricValue("total_restored_common_templates", operatorMetricsPort, &sspOperator)
+			}
+			return sum
+		}
+
+		saveOriginalVersion := func() {
+			originalVersion = ""
+
+			deployment := getDeployment(sspDeploymentName, sspDeploymentNamespace)
+			envs := deployment.Spec.Template.Spec.Containers[0].Env
+
+			for _, env := range envs {
+				if env.Name == common.OperatorVersionKey {
+					originalVersion = env.Value
+				}
+			}
+		}
+
+		updateVersion := func(version string) {
+			deployment := getDeployment(sspDeploymentName, sspDeploymentNamespace)
+
+			envs := deployment.Spec.Template.Spec.Containers[0].Env
+			var newEnvs []k8sv1.EnvVar
+
+			for _, env := range envs {
+				if env.Name != common.OperatorVersionKey {
+					newEnvs = append(newEnvs, env)
+				}
+			}
+
+			newEnvs = append(newEnvs, k8sv1.EnvVar{Name: common.OperatorVersionKey, Value: version})
+			deployment.Spec.Template.Spec.Containers[0].Env = newEnvs
+
+			Eventually(func() error {
+				return apiClient.Update(ctx, deployment)
+			}, 1*time.Minute, time.Second).ShouldNot(HaveOccurred())
+		}
+
+		waitForVersionUpdate := func(version string) {
+			Eventually(func() error {
+				operatorPods, _ := operatorPodsWithMetricsPort()
+
+				for _, pod := range operatorPods {
+					for _, container := range pod.Spec.Containers {
+						for _, env := range container.Env {
+							if env.Name == common.OperatorVersionKey && env.Value == version {
+								for _, condition := range pod.Status.Conditions {
+									if condition.Type == k8sv1.PodReady && condition.Status == k8sv1.ConditionTrue {
+										return nil
+									}
+								}
+							}
+						}
+					}
+				}
+
+				return errors.New("no pod has label " + common.OperatorVersionKey + " and is running")
+			}, 1*time.Minute, time.Second).ShouldNot(HaveOccurred())
+		}
+
+		BeforeEach(func() {
+			strategy.SkipSspUpdateTestsIfNeeded()
+			saveOriginalVersion()
+			waitUntilDeployed()
+		})
+
+		AfterEach(func() {
+			updateVersion(originalVersion)
+			waitForVersionUpdate(originalVersion)
+			strategy.RevertToOriginalSspCr()
+		})
+
+		It("[test_id:TODO]should not increment total_restored_common_templates during upgrades", func() {
```

Comment:
This test does not work as expected. I think these should be the steps:
- `pauseSsp()` - this pauses the reconciliation, and so any updated templates will not be restored
- Change the version environment variable on the SSP deployment - this simulates version update
- Update a template
- `unpauseSsp()` - when operator is unpaused it detects that the template was changed, but because it is in the middle of version update, it should not increase the metric.
- wait until SSP is in `deployed` state - then we know that version update is finished
- test that the metric was not increased.

Reply:
this is what is being done, except for the waiting ` until SSP is in deployed state` I think. Is this comment just to add that step?

Reply:
updated

Reply:
I meant that the test should look like this:
```go
It("[test_id:TODO]should not increment total_restored_common_templates during upgrades", func() {
	testTemplate := createTestTemplate()

	pauseSsp()
	version := "test-" + rand.String(5)
	updateVersion(version)
	waitForVersionUpdate(version)

	template := &templatev1.Template{}
	Expect(apiClient.Get(ctx, testTemplate.GetKey(), template)).To(Succeed())
	testTemplate.Update(template)
	Expect(apiClient.Update(ctx, template)).To(Succeed())

	unpauseSsp()
	waitUntilDeployed()

	newRestoredCount := totalRestoredTemplatesCount()
	Expect(newRestoredCount).To(BeZero())
})
```

The current test does not do what is expected.

Reply:
mhh, did't get it, updated